### ISSUE TITLE
ModelSnapshot: Ensure using directives present for annotation types

### ISSUE THREAD

**Issue by Bidou44** [aspnet/EntityFramework#2715]
I just installed the beta7-13841 and ran the `Script-Migration' command. It was working with the previous release, now I receive the following error: 

PM> Script-Migration
Specified cast is not valid.

Update: same behavior with Add-Migration

**Comment by tinganho** [comment]
I'm getting the error `The name 'SqlServerIdentityStrategy' does not exist in the current context`.

**Comment by NoralK** [comment]
Same here on beta7-12274
```
E:\Projects\HelloMvc\Migrations\PersonDBContextModelSnapshot.cs(16,66): DNX,Version=v4.5.1 error CS0103: The name 'SqlServerIdentityStrategy' does not exist in the current context
E:\Projects\HelloMvc\Migrations\PersonDBContextModelSnapshot.cs(22,74): DNX,Version=v4.5.1 error CS0103: The name 'SqlServerIdentityStrategy' does not exist in the current context
E:\Projects\HelloMvc\Migrations\20150726005437_MyFirstMigration.cs(17,74): DNX,Version=v4.5.1 error CS0103: The name 'SqlServerIdentityStrategy' doesnot exist in the current context
E:\Projects\HelloMvc\Migrations\20150726005437_MyFirstMigration.Designer.cs(26,66): DNX,Version=v4.5.1 error CS0103: The name 'SqlServerIdentityStrategy' does not exist in the current context
E:\Projects\HelloMvc\Migrations\20150726005437_MyFirstMigration.Designer.cs(32,74): DNX,Version=v4.5.1 error CS0103: The name 'SqlServerIdentityStrategy' does not exist in the current context
```

**Comment by branciat** [comment]
Also persist on dnx beta7-12275 with entityframework beta7-13844

**Comment by bricelam** [comment]
To clarify, you generated migration classes using an older version of EF (e.g. Beta 6) and you're trying to run `Script-Migration` on Beta 7?

If you delete your migrations (and model snapshot) and add a new one, does Script-Migration succeed?

**Comment by branciat** [comment]
I delete the migration files for each update of entity framework.
In the new migration files is missing an using on the assembly which include the SqlServerIdentityStrategy.

**Comment by NoralK** [comment]
+1 same for me

**Comment by NoralK** [comment]
With 7.0.0-beta7-13854 I cleared out all migrations and ran the following updated command:
```
dnx . ef migrations add MyFirstMigration --context PersonDBContext
```

I received these errors when compiling:
```
E:\Projects\HelloMvc\Migrations\PersonDBContextModelSnapshot.cs(16,66): DNX,Version=v4.5.1 error CS0103: The name 'SqlServerIdentityStrategy' does not exist in the current context
E:\Projects\HelloMvc\Migrations\PersonDBContextModelSnapshot.cs(22,74): DNX,Version=v4.5.1 error CS0103: The name 'SqlServerIdentityStrategy' does not exist in the current context
E:\Projects\HelloMvc\Migrations\20150730033506_MyFirstMigration.cs(17,74): DNX,Version=v4.5.1 error CS0103: The name 'SqlServerIdentityStrategy' doesnot exist in the current context
E:\Projects\HelloMvc\Migrations\20150730033506_MyFirstMigration.Designer.cs(26,66): DNX,Version=v4.5.1 error CS0103: The name 'SqlServerIdentityStrategy' does not exist in the current context
E:\Projects\HelloMvc\Migrations\20150730033506_MyFirstMigration.Designer.cs(32,74): DNX,Version=v4.5.1 error CS0103: The name 'SqlServerIdentityStrategy' does not exist in the current context
```
doing a little digging I found 'SqlServerIdentityStrategy' in Microsoft.Data.Entity.SqlServer.Metadata so I added this to all three migration files:
```
using Microsoft.Data.Entity.SqlServer.Metadata;
```
Now when I compile I am now get this error:
```
E:\Projects\HelloMvc\Migrations\20150730033506_MyFirstMigration.cs(18,74): DNX,Version=v4.5.1 error CS1503: Argument 2: cannot convert from 'Microsoft.Data.Entity.SqlServer.Metadata.SqlServerIdentityStrategy' to 'string'
```

Noral

**Comment by branciat** [comment]
Exactly, i have added the ToString() for each SqlServerIdentityStrategy.IdentityColumn because IdentityColumn is of type enum.

Compilation all right but when applied the migration file with dnx . ef migration apply initial i have the error below
```
dnx.exe : System.InvalidCastException: Le cast sp‚cifi‚ n'est pas valide.At line:1 char:1
+ dnx . ef migration apply initial
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (System.InvalidC...est pas valide.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
   … Microsoft.Data.Entity.SqlServer.SqlServerMigrationSqlGenerator.ColumnDefinition(String schema, String table, String name, String type, Boolean nullable, Object defaultValue, String defau
ltValueSql, String computedColumnSql, IAnnotatable annotatable, IModel model, SqlBatchBuilder builder)
   … Microsoft.Data.Entity.Migrations.Sql.MigrationSqlGenerator.ColumnDefinition(AddColumnOperation operation, IModel model, SqlBatchBuilder builder)
   … Microsoft.Data.Entity.Migrations.Sql.MigrationSqlGenerator.Generate(CreateTableOperation operation, IModel model, SqlBatchBuilder builder)
   … CallSite.Target(Closure , CallSite , Object , Object , IModel , SqlBatchBuilder )
   … System.Dynamic.UpdateDelegates.UpdateAndExecuteVoid4[T0,T1,T2,T3](CallSite site, T0 arg0, T1 arg1, T2 arg2, T3 arg3)
   … Microsoft.Data.Entity.Migrations.Sql.MigrationSqlGenerator.Generate(IReadOnlyList`1 operations, IModel model)
   … Microsoft.Data.Entity.Migrations.Migrator.ApplyMigration(Migration migration)
   … Microsoft.Data.Entity.Migrations.Migrator.ApplyMigrations(String targetMigration)
   … Microsoft.Data.Entity.Commands.MigrationTool.ApplyMigration(String migrationName, String contextTypeName, String startupAssemblyName)
   … Microsoft.Data.Entity.Commands.Program.<>c__DisplayClass13_0.<ApplyMigration>b__0()
   … Microsoft.Data.Entity.Commands.Program.Execute(String startupProject, Func`1 invoke)
   … Microsoft.Data.Entity.Commands.Program.ApplyMigration(String migration, String context, String startupProject)
   … Microsoft.Data.Entity.Commands.Program.<>c__DisplayClass10_3.<Main>b__9()
   … Microsoft.Framework.Runtime.Common.CommandLine.CommandLineApplication.Execute(String[] args)
   … Microsoft.Data.Entity.Commands.Program.Main(String[] args)
--- Fin de la trace de la pile … partir de l'emplacement pr‚c‚dent au niveau duquel l'exception a ‚t‚ lev‚e ---
   … System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   … Microsoft.Framework.Runtime.Common.EntryPointExecutor.Execute(Assembly assembly, String[] args, IServiceProvider serviceProvider)
   … Microsoft.Framework.ApplicationHost.Program.ExecuteMain(DefaultHost host, String applicationName, String[] args)
   … Microsoft.Framework.ApplicationHost.Program.Main(String[] args)
--- Fin de la trace de la pile … partir de l'emplacement pr‚c‚dent au niveau duquel l'exception a ‚t‚ lev‚e ---
   … System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   … Microsoft.Framework.Runtime.Common.EntryPointExecutor.Execute(Assembly assembly, String[] args, IServiceProvider serviceProvider)
   … dnx.host.Bootstrapper.RunAsync(List`1 args, IRuntimeEnvironment env, FrameworkName targetFramework)
   … dnx.host.RuntimeBootstrapper.ExecuteAsync(String[] args, FrameworkName targetFramework)
   … dnx.host.RuntimeBootstrapper.Execute(String[] args, FrameworkName targetFramework)
```

**Comment by NoralK** [comment]
@branciat I added the ToString() for the  ServerIdentityStrategy.IdentityColumn.  However, the command to apply the migration has changed.  The new command is:
```
dnx . ef database update --context PersonDBContext
```
the error I am getting now is:
```
E:\Projects\HelloMvc>dnx . ef database update --context PersonDBContext
Using context 'PersonDBContext'.
Using database 'aspnet5-BaseMVC-9adc29c4-988e-411c-96d3-15ad2e5a39e7' on server 'WIN81-ASPNET\SQLEXPRESS'.
System.InvalidCastException: Specified cast is not valid.
   at Microsoft.Data.Entity.SqlServer.SqlServerMigrationSqlGenerator.ColumnDefinition(String schema, String table, String name, String type, Boolean nullable, Object defaultValue, String defaultValueSql, String computedColumnSql, IAnnotatable annotatable, IModel model, SqlBatchBuilder builder)
   at Microsoft.Data.Entity.Migrations.Sql.MigrationSqlGenerator.ColumnDefinition(AddColumnOperation operation, IModel model, SqlBatchBuilder builder)
   at Microsoft.Data.Entity.Migrations.Sql.MigrationSqlGenerator.Generate(CreateTableOperation operation, IModel model, SqlBatchBuilder builder)
   at System.Dynamic.UpdateDelegates.UpdateAndExecuteVoid4[T0,T1,T2,T3](CallSite site, T0 arg0, T1 arg1, T2 arg2, T3 arg3)
   at Microsoft.Data.Entity.Migrations.Sql.MigrationSqlGenerator.Generate(IReadOnlyList`1 operations, IModel model)
   at Microsoft.Data.Entity.Migrations.Migrator.ApplyMigrations(String targetMigration)
   at Microsoft.Data.Entity.Commands.MigrationTool.ApplyMigration(String migrationName, String contextTypeName, String startupAssemblyName)
   at Microsoft.Data.Entity.Commands.Program.Execute(String startupProject, Action invoke)
   at Microsoft.Dnx.Runtime.Common.CommandLine.CommandLineUtilsExtensions.<>c__DisplayClass0_0.<OnExecute>b__0()
   at Microsoft.Dnx.Runtime.Common.CommandLine.CommandLineApplication.Execute(String[] args)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Dnx.Runtime.Common.EntryPointExecutor.Execute(Assembly assembly, String[] args, IServiceProvider serviceProvider)
   at Microsoft.Dnx.ApplicationHost.Program.ExecuteMain(DefaultHost host, String applicationName, String[] args)
   at Microsoft.Dnx.ApplicationHost.Program.Main(String[] args)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Dnx.Runtime.Common.EntryPointExecutor.Execute(Assembly assembly, String[] args, IServiceProvider serviceProvider)
   at Microsoft.Dnx.Host.Bootstrapper.RunAsync(List`1 args, IRuntimeEnvironment env, FrameworkName targetFramework)
   at Microsoft.Dnx.Host.RuntimeBootstrapper.ExecuteAsync(String[] args, FrameworkName targetFramework)
   at Microsoft.Dnx.Host.RuntimeBootstrapper.Execute(String[] args, FrameworkName targetFramework)
```

**Comment by bricelam** [comment]
The cast error is #2545

**Comment by smitpatel** [comment]
I investigated deeper in the issue.
It seems that with https://github.com/aspnet/EntityFramework/commit/baba8936d2db859f5f3fc2f95c72b1578b789239#diff-cd68493e81362e882059c78d0aa9ab01R288 commit we changed annotations in snapshot/migration files from string to enum type if any. We have not included using directive for that. This gives the error `The name 'SqlServerIdentityStrategy' does not exist in the current context`
This error occurs in generated Migration & ModelSnapshot files.
To fix this we can add using directive to the enum `SqlServerIdentityStrategy `.

Now the `OperationBuilder<TOperation>` expects Annotation value to be string. https://github.com/aspnet/EntityFramework/blob/dev/src/EntityFramework.Relational/Migrations/Builders/OperationBuilder.cs#L25. This gives error in migration file `Argument 2: cannot convert from 'Microsoft.Data.Entity.SqlServer.Metadata.SqlServerIdentityStrategy' to 'string'`
This error occurs in generated Migration files.
To fix this the enum `SqlServerIdentityStrategy` required to be casted to string using `ToString()` method.

The commit https://github.com/aspnet/EntityFramework/commit/ea658e95e47aebf2efd0889888e536bcc811458a#diff-4ae577e6726dfbb2a31b20bd07e1af96L271 casted the annotation value which is string type to enum `SqlServerIdentityStrategy` without using `TryParse`. This gives error `Specified cast is not valid.` This error throws at `Microsoft.Data.Entity.SqlServer.SqlServerMigrationSqlGenerator.ColumnDefinition`
